posit-dev/positron
# improve notebook interpreter start/stop ux

### Intent

Addresses #2561. This PR improves the UX around starting/stopping notebook interpreters, specifically:

1. Displays an error notification if starting a runtime fails.
2. The error notification has a "Retry" button.
3. If an interpreter failed to start, and the user tries to execute code, it will now restart the interpreter.

### Approach

In trying to do the above, I realized that the `notebookRuntimes` map was responsible for too many things and had become quite confusing, so I split that out into "starting" and "active" maps (roughly following `runtimeSession.ts`).

In my testing, it seems that we don't need to wait until the session is idle before trying to execute code, so I removed that. Can re-add if it becomes an issue.

These changes meant that the `NotebookRuntimeData` class was only tracking the execution order, so I brought that directly into the `NotebookController` and removed `NotebookRuntimeData`.

Given that it was already a big-ish diff, I renamed/reordered some things I'd been wanting to.

### QA Notes

See the linked issue for a repro. It's probably a good idea to play around with creating new notebooks, trying to execute code in them, and closing notebooks.
